contracts/payment/SponsorPayment.sol

```diff
@@ -0,0 +1,519 @@
+     * @param addr Address to check
+     */
+    function _isContract(address addr) private view returns (bool) {
+        uint256 size;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            size := extcodesize(addr)
+        }
+        return size > 0;
+    }
+
+    /**
+     * @dev Get the owner percentage value
+     * @return ownerPercentage
+     */
+    function getOwnerPercentage() external view returns (uint8) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        return $.ownerPercentage;
+    }
+
+    /**
+     * @dev Updates owner percentage value
+     * @param ownerPercentage Amount between 0 and 100 representing the owner percentage
+     */
+    function updateOwnerPercentage(
+        uint8 ownerPercentage
+    ) external onlyOwner validPercentValue(ownerPercentage) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        $.ownerPercentage = ownerPercentage;
+    }
+
+    function _tryRecoverSigner(
+        bytes32 digest,
+        bytes memory signature
+    ) private pure returns (address) {
+        (address signer, ECDSA.RecoverError err, ) = digest.tryRecover(signature);
+        if (err != ECDSA.RecoverError.NoError) revert InvalidPaymentClaim("Invalid signature");
+
+        return signer;
+    }
+
+    function recoverSponsorPaymentSigner(
+        SponsorPaymentInfo memory payment,
+        bytes memory signature
+    ) public view returns (address) {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    PAYMENT_CLAIM_DATA_TYPE_HASH,
+                    payment.recipient,
+                    payment.amount,
+                    payment.expiration,
+                    payment.nonce,
+                    keccak256(payment.metadata)
+                )
+            )
+        );
+
+        return _tryRecoverSigner(digest, signature);
+    }
+
+    function recoverSponsorPaymentSignerERC20(
+        ERC20SponsorPaymentInfo memory payment,
+        bytes memory signature
+    ) public view returns (address) {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    ERC20_PAYMENT_CLAIM_DATA_TYPE_HASH,
+                    payment.recipient,
+                    payment.amount,
+                    payment.token,
+                    payment.expiration,
+                    payment.nonce,
+                    keccak256(payment.metadata)
+                )
+            )
+        );
+
+        return _tryRecoverSigner(digest, signature);
+    }
+
+    /**
+     * @notice Deposits Ether or tokens as a sponsor.
+     */
+    function deposit() external payable nonReentrant {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+
+        if (msg.value == 0) revert InvalidDeposit("Invalid value amount");
+        $.balances[_msgSender()][address(0)] += msg.value;
```

Comment:
do we deposit to 0 address  for native token?

Reply:
Yes, I'm store native coin deposit per specific sponsor under address(0)
Any objection or suggestions?